# Task
Identify and display rows in the Excel file "/content/Copy of cf-products-barcodes (1).xlsx" that contain either duplicate or empty values in the 'fixed-barcode' column, and then summarize these findings.

## Load Excel File

### Subtask:
Load the Excel file into a pandas DataFrame.


**Reasoning**:
To load the Excel file and display its initial structure, I will import pandas, read the specified Excel file into a DataFrame named `df`, and then display the first few rows of the DataFrame.



In [1]:
import pandas as pd

# Load the Excel file into a pandas DataFrame
df = pd.read_excel('/content/Copy of cf-products-barcodes (1).xlsx')

# Display the first few rows of the DataFrame
df.head()

product_id            product_title    vendor product_type  \
0  8788552188053  Ombres G Eyeshadow Quad  Guerlain       Luxury   
1  8788552188053  Ombres G Eyeshadow Quad  Guerlain       Luxury   
2  8788552188053  Ombres G Eyeshadow Quad  Guerlain       Luxury   
3  8788552188053  Ombres G Eyeshadow Quad  Guerlain       Luxury   
4  8788552188053  Ombres G Eyeshadow Quad  Guerlain       Luxury   

                                              handle  \
0  https://cf.com.lb/products/ombres-g-eyeshadow-...   
1  https://cf.com.lb/products/ombres-g-eyeshadow-...   
2  https://cf.com.lb/products/ombres-g-eyeshadow-...   
3  https://cf.com.lb/products/ombres-g-eyeshadow-...   
4  https://cf.com.lb/products/ombres-g-eyeshadow-...   

                                         description      variant_id  \
0  Ombres G eyeshadow quads have an exceptional t...  46052464525461   
1  Ombres G eyeshadow quads have an exceptional t...  46052464558229   
2  Ombres G eyeshadow quads have an exceptional t...  46052464459925   
3  Ombres G eyeshadow quads have an exceptional t...  46052464590997   
4  Ombres G eyeshadow quads have an exceptional t...  46726528794773   

        variant_title      sku  price               color size  available  \
0   011 Imperial Moon  I045870  109.0   011 Imperial Moon  NaN       True   
1   214 Exotic Orchid  I045871  109.0   214 Exotic Orchid  NaN       True   
2      258 Wild Nudes  I049472  109.0      258 Wild Nudes  NaN       True   
3  360 Mystic Peacock  I045872  109.0  360 Mystic Peacock  NaN       True   
4    520 Stellar Glow  I056842  115.0    520 Stellar Glow  NaN      False   

   grams                                              image  fixed-barcode  
0      0  https://cdn.shopify.com/s/files/1/0716/5791/45...  3346470436558  
1      0  https://cdn.shopify.com/s/files/1/0716/5791/45...  3346470436510  
2      0  https://cdn.shopify.com/s/files/1/0716/5791/45...  3346470439634  
3      0  https://cdn.shopify.com/s/files/1/0716/5791/45...  3346470436503  
4      0  https://cdn.shopify.com/s/files/1/0716/5791/45...  3346470621909

## Identify Duplicate and Empty Fixed Barcodes

### Subtask:
Identify and filter all rows where the 'fixed-barcode' column has either duplicate values or is empty (NaN/null).


**Reasoning**:
To identify rows with empty or duplicate 'fixed-barcode' values, I will first create a boolean mask for empty values, then another for duplicate values (including first occurrences), combine these masks, and finally filter the DataFrame to create `df_issues`.



In [2]:
import numpy as np

# 1. Identify rows where the 'fixed-barcode' column is null or empty
# Convert potential empty strings to NaN first to catch all empty values
df['fixed-barcode'] = df['fixed-barcode'].replace('', np.nan)
empty_barcodes = df['fixed-barcode'].isnull()

# 2. Identify rows where the 'fixed-barcode' column contains duplicate values (including the first occurrence)
duplicate_barcodes = df['fixed-barcode'].duplicated(keep=False)

# 3. Combine these two conditions to create a single boolean mask
combined_mask = empty_barcodes | duplicate_barcodes

# 4. Apply this combined mask to filter and select rows that satisfy either condition
df_issues = df[combined_mask]

# Display the first few rows of the issues DataFrame and its shape
print("Rows with empty or duplicate 'fixed-barcode' values:")
print(df_issues.head())
print(f"\nTotal rows with issues: {len(df_issues)}")

Rows with empty or duplicate 'fixed-barcode' values:
       product_id       product_title               vendor product_type  \
16  8850359648405       Impact Parfum       Tommy Hilfiger       Luxury   
47  8984790597781          Cozy Nails      Samoa Cosmetics       Casual   
48  8984790597781          Cozy Nails      Samoa Cosmetics       Casual   
49  8984690229397  Lip Perfection Duo  Maybelline New York       Casual   
50  8984690229397  Lip Perfection Duo  Maybelline New York       Casual   

                                           handle  \
16       https://cf.com.lb/products/impact-parfum   
47          https://cf.com.lb/products/cozy-nails   
48          https://cf.com.lb/products/cozy-nails   
49  https://cf.com.lb/products/lip-perfection-duo   
50  https://cf.com.lb/products/lip-perfection-duo   

                                          description      variant_id  \
16  Impact Parfum  by Tommy Hilfiger is a Oriental...  46248382693525   
47  The Cozy Nails features: - 

**Reasoning**:
The current subtask to identify rows with empty or duplicate barcodes is complete. The next logical step is to summarize these findings as requested by the main task. This includes counting the number of empty barcodes, unique duplicate barcodes, and the total number of rows with issues.



In [3]:
empty_barcode_count = df_issues['fixed-barcode'].isnull().sum()
duplicate_barcode_count = df_issues['fixed-barcode'].duplicated(keep='first').sum()

# Count the number of unique duplicate barcode values (not including NaN)
unique_duplicate_values = df_issues['fixed-barcode'][df_issues['fixed-barcode'].duplicated(keep=False) & df_issues['fixed-barcode'].notnull()].nunique()

print("--- Summary of Fixed Barcode Issues ---")
print(f"Total rows with issues: {len(df_issues)}")
print(f"Number of rows with empty 'fixed-barcode': {empty_barcode_count}")
print(f"Number of unique duplicate 'fixed-barcode' values: {unique_duplicate_values}")
print("Note: A 'unique duplicate value' refers to a barcode string that appears more than once in the dataset.")


--- Summary of Fixed Barcode Issues ---
Total rows with issues: 985
Number of rows with empty 'fixed-barcode': 744
Number of unique duplicate 'fixed-barcode' values: 96
Note: A 'unique duplicate value' refers to a barcode string that appears more than once in the dataset.


## Display Filtered Rows

### Subtask:
Display the rows that have either duplicate or empty values in the 'fixed-barcode' column, categorizing them for clarity.


### Rows with Empty Barcodes

**Reasoning**:
Now that the markdown heading for empty barcodes is added, I will display the rows from `df_issues` that specifically have empty or null values in the 'fixed-barcode' column, showing only the relevant product identification and barcode columns as requested.



In [6]:
empty_barcode_rows = df_issues[df_issues['fixed-barcode'].isnull()]
print("Displaying rows with empty 'fixed-barcode' values:")
if not empty_barcode_rows.empty:
    print(empty_barcode_rows[['handle', 'product_title', 'fixed-barcode']].to_markdown(index=False))
else:
    print("No rows found with empty 'fixed-barcode' values.")

Displaying rows with empty 'fixed-barcode' values:
| handle                                                                                                         | product_title                                                                         |   fixed-barcode |
|:---------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------|----------------:|
| https://cf.com.lb/products/lip-perfection-duo                                                                  | Lip Perfection Duo                                                                    |             nan |
| https://cf.com.lb/products/lip-perfection-duo                                                                  | Lip Perfection Duo                                                                    |             nan |
| https://cf.com.lb/products/lip-perfection-duo                  

### Rows with Duplicate Barcodes

**Reasoning**:
Now that the markdown heading for duplicate barcodes is added, I will display the rows from `df_issues` that specifically have duplicate and non-empty 'fixed-barcode' values, showing only the relevant product identification and barcode columns as requested.



In [7]:
duplicate_non_empty_barcode_rows = df_issues[df_issues['fixed-barcode'].duplicated(keep=False) & df_issues['fixed-barcode'].notnull()]
print("Displaying rows with duplicate 'fixed-barcode' values (excluding empty values):")
if not duplicate_non_empty_barcode_rows.empty:
    print(duplicate_non_empty_barcode_rows[['handle', 'product_title', 'fixed-barcode']].to_markdown(index=False))
else:
    print("No rows found with duplicate 'fixed-barcode' values.")

Displaying rows with duplicate 'fixed-barcode' values (excluding empty values):
| handle                                                                                                                         | product_title                                                                                            |   fixed-barcode |
|:-------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------|----------------:|
| https://cf.com.lb/products/impact-parfum                                                                                       | Impact Parfum                                                                                            |   7640496672119 |
| https://cf.com.lb/products/cozy-nails                                                                                          | Cozy Nails           

In [11]:
urls = set(duplicate_non_empty_barcode_rows['handle'].tolist())
urls

{'https://cf.com.lb/products/addict-lip-glow',
 'https://cf.com.lb/products/all-of-me-intense-eau-de-parfum',
 'https://cf.com.lb/products/always-fabulous-airmat-foundation',
 'https://cf.com.lb/products/amber-musc-eau-de-parfum',
 'https://cf.com.lb/products/amore-mio-enchanted-aura',
 'https://cf.com.lb/products/amore-mio-soft-serve',
 'https://cf.com.lb/products/amore-mio-the-dark',
 'https://cf.com.lb/products/amore-mio-the-nude',
 'https://cf.com.lb/products/amore-mio-the-pastel',
 'https://cf.com.lb/products/amore-mio-the-pink',
 'https://cf.com.lb/products/amore-mio-the-red',
 'https://cf.com.lb/products/aqua-resist-brow-filler',
 'https://cf.com.lb/products/baked-illuminator',
 'https://cf.com.lb/products/bell-hypo-liquid-eye-liner',
 'https://cf.com.lb/products/besame-smoky-waterproof-kajal-eyeliner',
 'https://cf.com.lb/products/boi-ing-industrial-strength-concealer',
 'https://cf.com.lb/products/bottled-absolu-parfum-intense',
 'https://cf.com.lb/products/brow-styler',
 'htt

In [12]:
import requests
import re
import json
from typing import Optional

URLS = [
    'https://cf.com.lb/products/addict-lip-glow',
 'https://cf.com.lb/products/all-of-me-intense-eau-de-parfum',
 'https://cf.com.lb/products/always-fabulous-airmat-foundation',
 'https://cf.com.lb/products/amber-musc-eau-de-parfum',
 'https://cf.com.lb/products/amore-mio-enchanted-aura',
 'https://cf.com.lb/products/amore-mio-soft-serve',
 'https://cf.com.lb/products/amore-mio-the-dark',
 'https://cf.com.lb/products/amore-mio-the-nude',
 'https://cf.com.lb/products/amore-mio-the-pastel',
 'https://cf.com.lb/products/amore-mio-the-pink',
 'https://cf.com.lb/products/amore-mio-the-red',
 'https://cf.com.lb/products/aqua-resist-brow-filler',
 'https://cf.com.lb/products/baked-illuminator',
 'https://cf.com.lb/products/bell-hypo-liquid-eye-liner',
 'https://cf.com.lb/products/besame-smoky-waterproof-kajal-eyeliner',
 'https://cf.com.lb/products/boi-ing-industrial-strength-concealer',
 'https://cf.com.lb/products/bottled-absolu-parfum-intense',
 'https://cf.com.lb/products/brow-styler',
 'https://cf.com.lb/products/browza-eyebrow-compact-powder',
 'https://cf.com.lb/products/cafe-rose-eau-de-parfum',
 'https://cf.com.lb/products/capture-le-sa-c-rum-ox-c-treatment',
 'https://cf.com.lb/products/ck-be-eau-de-toilette',
 'https://cf.com.lb/products/cozy-nails',
 'https://cf.com.lb/products/dior-addict-shine-lipstick',
 'https://cf.com.lb/products/dior-forever-skin-glow-clean-radiant-24h-wear-foundation-spf20',
 'https://cf.com.lb/products/dior-sauvage-eau-de-parfum',
 'https://cf.com.lb/products/dior-sauvage-parfum',
 'https://cf.com.lb/products/diorshow-stylo-waterproof-eyeliner-24h-wear-intense',
 'https://cf.com.lb/products/diorsnow-uv-base-emulsion',
 'https://cf.com.lb/products/donna-eau-de-parfum',
 'https://cf.com.lb/products/double-serum-anti-aging-firming-serum',
 'https://cf.com.lb/products/dual-edge-eyeliner',
 'https://cf.com.lb/products/electro-touch-eau-de-parfum',
 'https://cf.com.lb/products/everlasting-youth-fluid-foundation',
 'https://cf.com.lb/products/fahrenheit-le-parfum-eau-de-parfum',
 'https://cf.com.lb/products/for-her-musc-nude-eau-de-parfum',
 'https://cf.com.lb/products/forever-better-skin-face-concealer',
 'https://cf.com.lb/products/forever-bond-base-coat',
 'https://cf.com.lb/products/forever-hydra-nude-foundation',
 'https://cf.com.lb/products/forever-nude-bronze',
 'https://cf.com.lb/products/forever-nude-bronzer-velvet-matte',
 'https://cf.com.lb/products/future-solution-lx-total-radiance-foundation',
 'https://cf.com.lb/products/gel-lip-liner-filler',
 'https://cf.com.lb/products/gel-nail-lacquer',
 'https://cf.com.lb/products/gift-card',
 'https://cf.com.lb/products/gimme-brow-volumizing-eyebrow-gel',
 'https://cf.com.lb/products/gimme-brow-volumizing-eyebrow-gel-mini',
 'https://cf.com.lb/products/glitter-bomb',
 'https://cf.com.lb/products/gloss-and-go-quick-dry',
 'https://cf.com.lb/products/glow-frost',
 'https://cf.com.lb/products/hoola-matte-bronzer-hoola-caramel',
 'https://cf.com.lb/products/hydra-essentiel-intensive-bi-phase-serum',
 'https://cf.com.lb/products/hypoallergenic-cover-eye-skin-concealer',
 'https://cf.com.lb/products/impact-parfum',
 'https://cf.com.lb/products/infaillible-32h-matte-cover-liquid-foundation-with-4-niacinamide-long-lasting-natural-finish-spf-25',
 'https://cf.com.lb/products/jeune-homme-eau-de-toilette',
 'https://cf.com.lb/products/joli-rouge-velvet-lipstick',
 'https://cf.com.lb/products/la-panthere-parfum',
 'https://cf.com.lb/products/lash-princess-false-lash-effect-mascara',
 'https://cf.com.lb/products/lifter-lip-liner',
 'https://cf.com.lb/products/lip-art-graphic-liner-liquid-lipstick',
 'https://cf.com.lb/products/love-me-eau-de-parfum',
 'https://cf.com.lb/products/love-me-the-emerald-elixir-eau-de-parfum',
 'https://cf.com.lb/products/love-me-the-onyx-eau-de-parfum',
 'https://cf.com.lb/products/love-me-the-silver-eau-de-parfum',
 'https://cf.com.lb/products/mademoiselle-jacadi-eau-de-toilette',
 'https://cf.com.lb/products/mademoiselle-petite-cerise-eau-de-toilette',
 'https://cf.com.lb/products/mademoiselle-petite-libellule-eau-de-toilette',
 'https://cf.com.lb/products/makeup-infaillible-32h-freshwear-longwear-liquid-foundation-vitamin-c-spf25',
 'https://cf.com.lb/products/matte-velvet-skin-full-coverage-foundation',
 'https://cf.com.lb/products/mens-impact-parfum',
 'https://cf.com.lb/products/modernmatte-powder-lipstick',
 'https://cf.com.lb/products/more-more-pink-eau-de-toilette',
 'https://cf.com.lb/products/never-nude-soft-serve',
 'https://cf.com.lb/products/never-nude-taste-of-winter',
 'https://cf.com.lb/products/never-nude-taste-of-winter-1',
 'https://cf.com.lb/products/never-nude-the-dark',
 'https://cf.com.lb/products/never-nude-the-neutral',
 'https://cf.com.lb/products/never-nude-the-nude',
 'https://cf.com.lb/products/never-nude-the-pastel',
 'https://cf.com.lb/products/never-nude-the-pink',
 'https://cf.com.lb/products/never-nude-the-red',
 'https://cf.com.lb/products/never-nude-the-white',
 'https://cf.com.lb/products/never-nude-tropical-breeze',
 'https://cf.com.lb/products/paris-cherry-glamour-eyeshadow-palette',
 'https://cf.com.lb/products/paris-evenings-palette',
 'https://cf.com.lb/products/pop-matte-lip-colour-primer',
 'https://cf.com.lb/products/power-of-plants-moisturising-face-gel-aloe',
 'https://cf.com.lb/products/professional-wooden-nail-file',
 'https://cf.com.lb/products/pure-color-envy-luxe-eyeshadow-quad',
 'https://cf.com.lb/products/quickliner-for-eyes-eyeliner',
 'https://cf.com.lb/products/radiant-lift-foundation',
 'https://cf.com.lb/products/rose-musc-eau-de-parfum',
 'https://cf.com.lb/products/sauvage-eau-forte-parfum',
 'https://cf.com.lb/products/shiseido-modern-matte-powder-lipstick',
 'https://cf.com.lb/products/skinshield-creamy-foundation-spf30',
 'https://cf.com.lb/products/skinshield-foundation-spf30-bundle',
 'https://cf.com.lb/products/soft-touch-mousse-blush',
 'https://cf.com.lb/products/sorbet-garden-eau-de-toilette',
 'https://cf.com.lb/products/sound-of-donna-eau-de-parfum',
 'https://cf.com.lb/products/sports-bb-spf-50-very-water-resistant',
 'https://cf.com.lb/products/synchro-skin-self-refreshing-foundation-spf30',
 'https://cf.com.lb/products/synchro-skin-self-refreshing-tint-spf20',
 'https://cf.com.lb/products/the-eye-pencil',
 'https://cf.com.lb/products/touch-the-luminous-gold-eau-de-toilette',
 'https://cf.com.lb/products/tous-eau-de-parfum',
 'https://cf.com.lb/products/tous-eau-de-toilette',
 'https://cf.com.lb/products/tous-man-eau-de-toilette',
 'https://cf.com.lb/products/tous-man-sport-eau-de-toilette',
 'https://cf.com.lb/products/tous-man-spritz-eau-de-toilette',
 'https://cf.com.lb/products/tout-petit-eau-de-senteur',
 'https://cf.com.lb/products/toute-petite-eau-de-senteur',
 'https://cf.com.lb/products/ultra-hd-invisible-cover-foundation',
 'https://cf.com.lb/products/velvet-finish-powder-compact',
 'https://cf.com.lb/products/volume-glamour-eyeshadow-palette',
 'https://cf.com.lb/products/wanted-by-night-eau-de-parfum',
 'https://cf.com.lb/products/waterproof-eye-pencil',
 'https://cf.com.lb/products/your-moments-eau-de-toilette',
 'https://cf.com.lb/products/your-powers-eau-de-toilette'
]

HEADERS = {
    "User-Agent": "Mozilla/5.0 (compatible; ProductScraper/1.0)"
}

SEEDGROW_REGEX = re.compile(
    r'window\.seedgrow_current_product\s*=\s*({.*?})\s*;',
    re.DOTALL
)


def extract_seedgrow_json(html: str) -> Optional[dict]:
    match = SEEDGROW_REGEX.search(html)
    if not match:
        return None
    return json.loads(match.group(1))


results = []
failed = []

for idx, url in enumerate(URLS, start=1):
    try:
        response = requests.get(url, headers=HEADERS, timeout=15)
        response.raise_for_status()

        product_json = extract_seedgrow_json(response.text)

        if product_json:
            results.append(product_json)
            print(f"{idx}. OK")
        else:
            failed.append(url)
            print(f"{idx}. NO_JSON")

    except Exception:
        failed.append(url)
        print(f"{idx}. ERROR")


# -------------------------
# SAVE TO FILES
# -------------------------
with open("products_with_duplicated_barcodes.json", "w", encoding="utf-8") as f:
    json.dump(results, f, ensure_ascii=False, indent=2)

with open("failed_urls.json", "w", encoding="utf-8") as f:
    json.dump(failed, f, ensure_ascii=False, indent=2)


print("\n✅ Finished")
print(f"Saved products: {len(results)}")
print(f"Failed URLs: {len(failed)}")

1. OK
2. OK
3. OK
4. OK
5. OK
6. OK
7. OK
8. OK
9. OK
10. OK
11. OK
12. OK
13. OK
14. OK
15. OK
16. OK
17. OK
18. OK
19. OK
20. OK
21. OK
22. OK
23. OK
24. OK
25. OK
26. OK
27. OK
28. OK
29. OK
30. OK
31. OK
32. OK
33. OK
34. OK
35. OK
36. OK
37. OK
38. OK
39. OK
40. OK
41. OK
42. OK
43. OK
44. OK
45. OK
46. OK
47. OK
48. OK
49. OK
50. OK
51. OK
52. OK
53. OK
54. OK
55. OK
56. OK
57. OK
58. OK
59. OK
60. OK
61. OK
62. OK
63. OK
64. OK
65. OK
66. OK
67. OK
68. OK
69. OK
70. OK
71. OK
72. OK
73. OK
74. OK
75. OK
76. OK
77. OK
78. OK
79. OK
80. OK
81. OK
82. OK
83. OK
84. OK
85. OK
86. OK
87. OK
88. OK
89. OK
90. OK
91. OK
92. OK
93. OK
94. OK
95. OK
96. OK
97. OK
98. OK
99. OK
100. OK
101. OK
102. OK
103. OK
104. OK
105. OK
106. OK
107. OK
108. OK
109. OK
110. OK
111. OK
112. OK
113. OK
114. OK
115. OK
116. OK
117. OK
118. OK
119. OK

✅ Finished
Saved products: 119
Failed URLs: 0


## Summary:

### Data Analysis Key Findings

*   **Total Issues Identified**: A total of 985 rows were found to have issues in the 'fixed-barcode' column.
*   **Empty Barcodes**: 744 rows have an empty or null value in the 'fixed-barcode' column, indicating a significant number of products are missing barcode information.
*   **Duplicate Barcode Values**: 96 unique 'fixed-barcode' values are duplicated across different product entries. This means the same barcode is assigned to multiple distinct products.
    *   For example, barcode `7640496672119` is assigned to both 'Impact Parfum' (product\_id 8850359648405) and 'Men's Impact Parfum' (product\_id 8966359122069).
    *   Other notable duplicate barcodes include `5281011237301`, `4059729393876`, `3380810318692`, `4059729348845`, `20714833008`, `5281011177003`, `5281011181000`, and `3348901729208`, each appearing multiple times across different products.

### Insights or Next Steps

*   **Data Quality Improvement**: The identified empty and duplicate barcodes indicate critical data quality issues that need to be addressed to ensure accurate product identification and inventory management.
*   **Root Cause Analysis**: Investigate the source of these data inconsistencies. This could involve reviewing data entry processes, understanding system integrations, or identifying potential data migration errors to prevent future occurrences.
